plat/hikey/plat_pm.c

```diff
@@ -207,10 +207,24 @@ static void hikey_affinst_suspend_finish(uint32_t afflvl,
 
 static void __dead2 hikey_system_off(void)
 {
-	gpio_set_value(0, 0);
-	mdelay(1000);
-	/* Send the system reset request since it fails to power off */
-	mmio_write_32(AO_SC_SYS_STAT0, 0x48698284);
+	NOTICE("%s: off system\n", __func__);
+
+	/* pulling GPIO_0_0 low to trigger PMIC shutdown */
+	/* setting pinmux */
+	mmio_write_32(0xF8001810, 0x2);
+	/* setting pin direction */
+	mmio_write_8(0xF8011400, 1);
+	/* setting pin output value */
+	mmio_write_8(0xF8011004, 0);
+
+	/* PMIC shutdown depends on two conditions: GPIO_0_0 (PWR_HOLD) low,
+	 * and VBUS_DET < 3.6V. For HiKey, VBUS_DET is connected to VDD_4V2
+	 * through Jumper 1-2. So, to complete shutdown, user needs to manually
+	 * remove Jumper 1-2.
+	 */
+	NOTICE(".........................................................\n");
+	NOTICE(" IMPORTANT: Remove Jumper 1-2 to complete shutdown\n");
+	NOTICE(" DANGER:    SoC is still burning. DANGER!\n");
```

Comment:
On 15 June 2016 at 07:34, John Stultz <notifications@github.com> wrote:

> In plat/hikey/plat_pm.c
> <https://github.com/96boards-hikey/arm-trusted-firmware/pull/3#discussion_r67077109>
> :
>
> > +	/* pulling GPIO_0_0 low to trigger PMIC shutdown */
> > +	/* setting pinmux */
> > +	mmio_write_32(0xF8001810, 0x2);
> > +	/* setting pin direction */
> > +	mmio_write_8(0xF8011400, 1);
> > +	/* setting pin output value */
> > +	mmio_write_8(0xF8011004, 0);
> > +
> > +	/* PMIC shutdown depends on two conditions: GPIO_0_0 (PWR_HOLD) low,
> > +	 * and VBUS_DET < 3.6V. For HiKey, VBUS_DET is connected to VDD_4V2
> > +	 * through Jumper 1-2. So, to complete shutdown, user needs to manually
> > +	 * remove Jumper 1-2.
> > +	 */
> > +	NOTICE(".........................................................\n");
> > +	NOTICE(" IMPORTANT: Remove Jumper 1-2 to complete shutdown\n");
> > +	NOTICE(" DANGER:    SoC is still burning. DANGER!\n");
>
> This doesn't seem safe. Should we instead reboot the machine if the jumper
> is present?
>

Reboot also has its problem. The user wants to shutdown, but he finds it
always 'fails'.

BTW, we also need a balance here between effort and result. If I have
enough time to work on this, I will agree with you. It's better to check
the Jumper status, and do something like blinking all 4 LEDs to catch
user's attention, then if still fails, I can do a reset, and in the reset,
only start one core with minimum kernel.

Or, before the kernel falls into PSCI system-off, it can check the Jumper
status. If it's connected, kernel should never shutdown...

-Guodong


—
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/96boards-hikey/arm-trusted-firmware/pull/3/files/08515f4ed85f9e345120dec07e742cb36799b05d#r67077109>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AHnKpyNACAbLnMa1Tis2aYz03izZ100Xks5qLzqSgaJpZM4I0_Ji>
> .
>